src/awkward/contents/listoffsetarray.py

```diff
@@ -1986,10 +1986,73 @@ def _to_arrow(self, pyarrow, mask_node, validbytes, length, options):
 
     def _to_backend_array(self, allow_missing, backend):
         array_param = self.parameter("__array__")
-        if array_param in {"bytestring", "string"}:
-            # As our array-of-strings _may_ be empty, we should pass the dtype
-            dtype = np.str_ if array_param == "string" else np.bytes_
-            return backend.nplike.asarray(self.to_list(), dtype=dtype)
+        if array_param == "string":
+            # Determine the widest string (in code points)
+            _max_code_points = backend.index_nplike.empty(1, dtype=np.int64)
+            backend[
+                "awkward_NumpyArray_prepare_utf8_to_utf32_padded",
+                self._content.dtype.type,
+                self._offsets.dtype.type,
+                _max_code_points.dtype.type,
+            ](
+                self._content.data,
+                self._offsets.data,
+                self._offsets.length,
+                _max_code_points,
+            )
+            max_code_points = backend.index_nplike.index_as_shape_item(
+                _max_code_points[0]
+            )
+            # Ensure that we have at-least length-1 bytestrings
+            if max_code_points is not unknown_length:
+                max_code_points = max(1, max_code_points)
+
+            # Allocate the correct size buffer
+            total_code_points = max_code_points * self.length
+            buffer = backend.nplike.empty(total_code_points, dtype=np.uint32)
```

Comment:
According to the comments and function names, this first step finds the maximum number of code points of any string in the array and uses that to allocate the buffer into which UTF-32 data will be written.

~~Shouldn't it be finding the maximum number of _bytes_, not _code points_? The size of the buffer into which UTF-32 data will be written is specified as a number of bytes.~~

Nope! I'm wrong! In UTF-32, each code point is a 4-byte integer, all the same size.

```python
buffer = backend.nplike.empty(total_code_points, dtype=np.uint32)
```

makes a `buffer` that is the correct number of bytes for UTF-32.